### ISSUE TITLE
Rust Bindings

### ISSUE THREAD

**Issue by callum-gander** [electric-sql/pglite#470]
Are there any future plans to add Rust bindings for this? I can see this as an appealing alternative to bundling up Postgres as a sidecar for things like Tauri apps or using SQLite. Thanks

**Comment by realmorrisliu** [comment]
> Are there any future plans to add Rust bindings for this? I can see this as an appealing alternative to bundling up Postgres as a sidecar for things like Tauri apps or using SQLite. Thanks

Maybe you could try using pglite directly within Tauri’s webview? I’m not sure if it will work, but it could be an interesting alternative to bundling Postgres as a sidecar.

**Comment by callum-gander** [comment]
That doesn't feel very secure imo, feels like it should sit on the rust side rather than in the webview, but appreciate the idea

**Comment by pmp-p** [comment]
Not keen on Rust, but maybe a valid approach would be to just run pglite inside rust with https://docs.wasmtime.dev/lang-rust.html like it was done for Go here https://github.com/electric-sql/pglite/issues/89

To go further than that POC you would need to pick a rust postgres wire protocol client crate though.

**Comment by dlight** [comment]
Besides wasm, can pglite also build to native code? Getting Rust to run regular wasm pglite would be awesome, but the downside is that it would still require a wasm runtime. Embedding a native build of pglite as a library in a Rust program would be very useful, if at all possible.

Right now what exists in Rust for "embedding" Postgres is [pg_embed](https://github.com/faokunega/pg-embed) and [postgresql-embedded](https://github.com/theseus-rs/postgresql-embedded). They can use native binaries from [embedded-postgres-binaries](https://github.com/zonkyio/embedded-postgres-binaries) (which is a trimmed down Postgres made for testing SQL queries in the Java ecosystem), or, in the case of `postgresql-embedded`, also from [postgresql-binaries](https://github.com/theseus-rs/postgresql-binaries). `pg_embed` downloads the Postgres binary in the first run, and `postgresql-embedded` can also do that, but can package the Postgres binary inside its own binary to make distribution easier, extract it somewhere in the first run either way - Postgres still runs as a separate program. An upside for `postgresql-embedded` is that it supports Postgres extensions though.

**Comment by tdrz** [comment]
> Besides wasm, can pglite also build to native code? 

Not yet, but we'd like it to. See the last part of [this comment](https://github.com/electric-sql/pglite/issues/358#issuecomment-2381408186).

**Comment by realmorrisliu** [comment]
Super excited for `libpglite` to come!

**Comment by evanlhatch** [comment]
> > Besides wasm, can pglite also build to native code?
> 
> Not yet, but we'd like it to. See the last part of [this comment](https://github.com/electric-sql/pglite/issues/358#issuecomment-2381408186).

native rust would be huge for me- any idea on when this will be coming?

**Comment by pmp-p** [comment]
i think the correct way -now- would be to embded wasmtime. Only wasm sandboxing can offer safety that would be lost by compiling pglite as transpiled rust and/or unsafe C library.

Wasmtime embedding is quick, easy and ultra portable. Any other mean would probably be only possible as a long term project.

**Comment by paulm17** [comment]
Why is wasmtime the way to go?  

> We want to extract the core changes to Postgres that we had to make and create a "libpglite" C lib that can be linked to from any language/platform.

Taken from: https://news.ycombinator.com/item?id=41225185

Wouldn't a libpglite be the most preferable way?

> unsafe C library.

Wouldn't any issues be handled the rust side?

**Comment by TheKnarf** [comment]
I'd love a working example to work off on if anyone has gotten wasmtime + pglite wasm file working.

**Comment by f0rr0** [comment]
I was able to get some basic stuff working for rust and wasmtime here: [pglite-oxide](https://github.com/f0rr0/pglite-oxide)

**Comment by yujonglee** [comment]
@f0rr0 super cool. what are current limitations?

**Comment by f0rr0** [comment]
@yujonglee you can run a unix/tcp socket and talk over wire protocol. or you can use the CMA channel and read/write command inputs/outputs. I also include the vector extension and pgdump. I am currently using it in my tauri desktop app so I guess I will eventually find out the limitations.

**Comment by paulm17** [comment]
@f0rr0 Thanks for putting this together.  Would be nice if the ReadMe could have a section for Tauri integration!  😄

**Comment by f0rr0** [comment]
Yes I’ll be adding that soon once I successfully integrate and run the
whole thing

**Comment by roo12312** [comment]
any updates on how to use with Tauri?

**Comment by Nicolab** [comment]
I migrated my Tauri backend from a Node sidecar to Rust, so I had to switch from pglite to SQLite—ouch. It would be awesome to have pglite support with SQLx (might as well use a battle-tested and widely adopted library).
Having a support in Rust would open the door to other languages through FFI.

**Comment by f0rr0** [comment]
I’ve been a bit busy but will do an example tomorrow. I already use this is
in my tauri app

**Comment by kshcherban** [comment]
I also did it [here](https://github.com/kshcherban/pglite-rust-bindings), my main purpose was to use it in tests (tokio::test) instead of testcontainers. Mostly done with AI but tested and works for me. First run is slow (~1m as it JIT compiles wasi runtime), then it uses cached files and is almost instant (postgres engine starts in ~2s).

**Comment by luke-stdev001** [comment]
I've just listed my own use case and watching the incredible work being done in: [#842](https://github.com/electric-sql/pglite/pull/842) (libpglite).

**My Use Case**
I'm building an offline-first app with Tauri 2.x targeting iOS, Android, Windows, and Linux. The architecture has:

Business logic in Rust (not just the webview JS layer)
PostgreSQL 18 on the server
Need for PostgreSQL semantics locally (UUIDv7, JSONB, proper transactions)
Sync via Durable Streams between client and server

**Currently, my options are:**

**SQLite** - Already Works and I have this in production, but loses PostgreSQL feature parity with the server, i'm reaching for Postgres-backed BM25 search, native UUIDv7 indexes, pg_trgm, and full type safety between fat client and thin server both running Postgres 18.0.
**Embedded Postgres sidecar** - Heavy, complex deployment on mobile
**PGLite in webview only** - Can't access from Rust layer where my business logic lives

**What I'd Need**
Rust bindings (via C FFI from libpglite) that let me do something like:

```
use pglite::PgLite;

let db = PgLite::new("./data/local.db")?;
db.exec("CREATE TABLE IF NOT EXISTS events (id UUID PRIMARY KEY DEFAULT uuidv7(), ...)")?;

let rows = db.query("SELECT * FROM events WHERE id = $1", &[&event_id])?;
```

This would allow the same sqlx patterns I use on the server to work locally, with PGLite as the backing store.
**Re: PR #842**
The libpglite work by @tdrz looks like exactly the foundation needed for this. Once there's a stable C API, Rust bindings via bindgen or a hand-written FFI wrapper would be straightforward.
has anyone played with libpglite in it's current state, or perhaps know of any internal roadmap? @tdrz , i'd be happy to help test on Tauri/mobile targets if believe's it's at a stage it's ready to play with.

**Comment by tdrz** [comment]
> [#842](https://github.com/electric-sql/pglite/pull/842) (libpglite)

I had to switch to other tasks but I'm picking this up again.

> [@tdrz](https://github.com/tdrz) , i'd be happy to help test on Tauri/mobile targets if believe's it's at a stage it's ready to play with.

Thank you, please keep an eye on the project!

**Comment by satvikpendem** [comment]
@luke-stdev001 I'm doing the same thing, local first app using Rust on client and server, what did you end up using a month later? Or are you still waiting on libpglite?

**Comment by f0rr0** [comment]
Was quite busy and took me a while to get around it but I just released `0.2.0` of `pglite-oxide`. Among other things I added:

- compatibility coverage for `sqlx` and `tokio-postgres`
- Tauri usage docs 
- a lot more tests and high level APIs

Check it out: https://github.com/f0rr0/pglite-oxide/releases/tag/0.2.0
